### PR TITLE
Dash to effect

### PR DESCRIPTION
**Pull request by nick-wilson95**

### CODE REVIEW

**Comment by nick-wilson95** [comment]
[Move dash to an effect](https://trello.com/c/R38Nx0kH/16-move-dash-to-an-effect)

File: scripts/MACROS_oPlayer/MACROS_oPlayer.gml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// Dash config
+#macro PLAYER_DASH_SPEED 3.3
+#macro PLAYER_DASH_DURATION 6
+
+#macro PLAYER_DASH_SLOW 0.25
+#macro PLAYER_DASH_SLOW_DURATION 20
+
+#macro PLAYER_DASH_COOLDOWN 60
```
Comment (Drummerboy444) [inline scripts/MACROS_oPlayer/MACROS_oPlayer.gml]: I think it feels a bit weird that there's a period of time where you're not slowed but you can't dash. I think the dash cooldown should probably be the sum of the dash duration and the dash slow duration

File: scripts/Player_Move/Player_Move.gml [hunk 2]
```diff
@@ -26,51 +24,36 @@ var _num_movementSpeed = ds_Stats_Get(oPlayer.ds_Stats_frameStats, Enum_Stats.SP
...
-				num_slowTimer = 0;
-				enum_currentMoveMode = Enum_PlayerMoveModes.FREEMOVE;
-			} else {
-				num_slowTimer++;
-			}
-			break;	
-		#endregion
+	
+	if(AbilityCaster_CheckInstanceHasEffect(id, Enum_Effects.DASH)){
+		MovementData_Movement = MovementData_New(
```
Comment (Drummerboy444) [inline scripts/Player_Move/Player_Move.gml]: `MovementData_Movement` shouldn't really be an instance variable. I think the nicest thing to do here would be to have 2 local variables declared before this if block, say `_num_xOffset` and `_num_yOffset` that are just set to 0. Then inside the if else blocks we just set those accordingly. Then the line `MovementManager_QueueMovement(MovementData_Movement);` could become something like:
```js
MovementManager_QueueMovement(
	MovementData_New(
		x + _num_xOffset,
		y + _num_yOffset,
		id,
		true
	)
);
```

File: scripts/Dash_BucketHandler/Dash_BucketHandler.gml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// There should be at most one dash effect on the player, so we just deal with the first element in the list.
+if(ds_list_size(_list_bucket)==0) return;
+
+var _EffectData_data = _list_bucket[| 0];
+var _arr_arguments = EffectData_GetArguments(_EffectData_data);
+var _num_CurrentSpeed = ds_Stats_Get(_id_abilityCaster.ds_Stats_frameStats, Enum_Stats.SPEED);
+	
+ds_Stats_Set(_id_abilityCaster.ds_Stats_frameStats, Enum_Stats.SPEED, _num_CurrentSpeed * _arr_arguments[0]);
+
+if(EffectData_GetSteps(_EffectData_data) = 0){
```
Comment (Drummerboy444) [inline scripts/Dash_BucketHandler/Dash_BucketHandler.gml]: I think `<= 0` would be good here, just for safety

File: scripts/Dash_BucketHandler/Dash_BucketHandler.gml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+	AbilityCaster_AddEffectDataToBuffer(
+		_id_abilityCaster,
+		EffectData_New(Enum_Effects.SLOW, [_arr_arguments[1]], _arr_arguments[2], _id_abilityCaster)
```
[the remark this answers is not in the document]
Comment (nick-wilson95) [reply scripts/Dash_BucketHandler/Dash_BucketHandler.gml]: Yeah I realised our current system is pretty unsustainable - I got pretty bamboozled about the dash effect data and only needed it in 2 places.

I like the inheritance framework for different varieties of effect data. I'm pretty sure it doesn't belong in this branch though so let's raise it in the next formation lap.

[the remark this answers is not in the document]
Comment (Drummerboy444) [reply scripts/Dash_BucketHandler/Dash_BucketHandler.gml]: Yeah I think you're right, it sounds like another task to me. I'll hold off on merging this until we've had a chat about all that stuff on Wednesday

**Pull request merged**